examples/qualcomm/llama2/qaihub_runner/io_memory.cpp

```diff
@@ -0,0 +1,477 @@
+        const_cast<TensorImpl::SizesType*>(input_ids->sizes().data()),
+        ptr->input_ids,
+        const_cast<TensorImpl::DimOrderType*>(input_ids->dim_order().data()));
+    input_tensors_[0].push_back(input_ids_.get());
+    // [I]: atten_mask
+    Result<TensorInfo> atten_mask = methods_meta[0]->input_tensor_meta(1);
+    attention_mask_ = std::make_unique<TensorImpl>(
+        atten_mask->scalar_type(),
+        atten_mask->sizes().size(),
+        const_cast<TensorImpl::SizesType*>(atten_mask->sizes().data()),
+        ptr->attention_mask,
+        const_cast<TensorImpl::DimOrderType*>(atten_mask->dim_order().data()));
+    input_tensors_[0].push_back(attention_mask_.get());
+    // [I]: pos_ids_cos
+    Result<TensorInfo> pos_ids_cos = methods_meta[0]->input_tensor_meta(2);
+    position_ids_cos_ = std::make_unique<TensorImpl>(
+        pos_ids_cos->scalar_type(),
+        pos_ids_cos->sizes().size(),
+        const_cast<TensorImpl::SizesType*>(pos_ids_cos->sizes().data()),
+        ptr->position_ids_cos,
+        const_cast<TensorImpl::DimOrderType*>(pos_ids_cos->dim_order().data()));
+    input_tensors_[0].push_back(position_ids_cos_.get());
+    // [I]: pos_ids_sin
+    Result<TensorInfo> pos_ids_sin = methods_meta[0]->input_tensor_meta(3);
+    position_ids_sin_ = std::make_unique<TensorImpl>(
+        pos_ids_sin->scalar_type(),
+        pos_ids_sin->sizes().size(),
+        const_cast<TensorImpl::SizesType*>(pos_ids_sin->sizes().data()),
+        ptr->position_ids_sin,
+        const_cast<TensorImpl::DimOrderType*>(pos_ids_sin->dim_order().data()));
+    input_tensors_[0].push_back(position_ids_sin_.get());
+    // [IO]: hidden_state => [I] shard2,3,4
+    int output_index = 8 * 2 * 32; // layres*(k + v caches)*heads
+    Result<TensorInfo> hidden_state =
+        methods_meta[0]->output_tensor_meta(output_index);
+    hidden_state_ = std::make_unique<TensorImpl>(
+        hidden_state->scalar_type(),
+        hidden_state->sizes().size(),
+        const_cast<TensorImpl::SizesType*>(hidden_state->sizes().data()),
+        ptr->hidden_state,
+        const_cast<TensorImpl::DimOrderType*>(
+            hidden_state->dim_order().data()));
+    // reuse inputs for following tensors
+    for (int shard_index = 1; shard_index < 4; ++shard_index) {
+      // inpus of shard1,2,3: hidden_state, atten_mask, pos_ids_cos, pos_ids_sin
+      input_tensors_[shard_index].push_back(hidden_state_.get());
+      input_tensors_[shard_index].push_back(attention_mask_.get());
+      input_tensors_[shard_index].push_back(position_ids_cos_.get());
+      input_tensors_[shard_index].push_back(position_ids_sin_.get());
+    }
+  }
+  // [O] kv_cache for shard1,2,3,4
+  for (int offset = 0, shard_index = 0; shard_index < 4;
+       offset += 8, shard_index++) {
+    for (int i = 0; i < 8; ++i) { // layers per shard
+      for (int j = 0; j < 2; ++j) { // k_cache + v_cache
+        for (int k = 0; k < 32; ++k) { // heads
```

Comment:
might as well make some named constants for these and then you don't need the comments here and below

Reply:
Changed, thanks.